Sources/swift-format/Frontend/Frontend.swift

```diff
@@ -124,27 +124,42 @@ class Frontend {
   }
 
   /// Processes source content from a list of files and/or directories provided as paths.
-  private func processPaths(_ paths: [String]) {
+  private func processPaths(_ paths: [String], parallel: Bool) {
     precondition(
       !paths.isEmpty,
       "processPaths(_:) should only be called when paths is non-empty.")
 
-    for path in FileIterator(paths: paths) {
+    let concurrentQueue = DispatchQueue(
+        label: "swift-format-queue", qos: .userInitiated, attributes: .concurrent)
+    let lock = NSLock()
+    let openAndProcess: (String) -> Void = { path in
       guard let sourceFile = FileHandle(forReadingAtPath: path) else {
-        diagnosticEngine.diagnose(Diagnostic.Message(.error, "Unable to open \(path)"))
-        continue
+        lock.lock()
+        defer { lock.unlock() }
+        self.diagnosticEngine.diagnose(Diagnostic.Message(.error, "Unable to open \(path)"))
+        return
       }
 
-      guard let configuration = configuration(
-        atPath: lintFormatOptions.configurationPath, orInferredFromSwiftFileAtPath: path)
+      guard let configuration = self.configuration(
+        atPath: self.lintFormatOptions.configurationPath, orInferredFromSwiftFileAtPath: path)
       else {
         // Already diagnosed in the called method.
-        continue
+        return
       }
 
       let fileToProcess = FileToProcess(
         fileHandle: sourceFile, path: path, configuration: configuration)
-      processFile(fileToProcess)
+      self.processFile(fileToProcess)
+    }
+
+    for path in FileIterator(paths: paths) {
+      if parallel {
+        concurrentQueue.async {
+          openAndProcess(path)
+        }
+      } else {
+        openAndProcess(path)
+      }
```

Comment:
You'll still need a `DispatchGroup` to wait on here, right? Otherwise the process could exit after enqueuing but not necessarily exiting all of the work.